homeassistant/components/auth/__init__.py

```diff
@@ -182,6 +183,19 @@ async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
 
     hass.data[DOMAIN] = store_result
 
+    async def _check_expired_refresh_tokens(_: datetime | None = None) -> None:
+        """Check for and delete expired refresh tokens."""
+        await hass.auth.async_remove_expired_tokens()
+
+    await _check_expired_refresh_tokens()
+    async_track_time_interval(
+        hass,
+        _check_expired_refresh_tokens,
+        timedelta(seconds=60),
```

Comment:
```suggestion
        timedelta(hours=1),
```
You are going to expire a token that lived for 90 days without usage, I think this check can be more relaxed to do it once every hour (even every 10 minutes would be better)

Reply:
Since we know when the next one expires, we could schedule the timer for than instead and schedule the next time for the next one so they all expire exactly on time and this check only runs when we know one is ready to go

Reply:
This was also my first idea, but the expiration dates can change, as soon as a token is used, so you track all of these scheduled tasks and replace/recreate/update them, as soon as the related token is used 🤔

Reply:
after some ☕ I think I've an idea how to do this ...

Reply:
it has been reworked, so that the `homeassistant.component.auth` is not involved anymore, but everything is handled within the `AuthManager` itself.
- during init of the `AuthManager` tasks for each token with an expiration will be scheduled
- when creating a new refresh token, a task will be scheduled (_if needed_)
- when a refresh token get's used, the former expiration task will be canceled and a new one is scheduled

Reply:
I think that's overkill. I would just create a timer for the first one to expire. Generally in a healthy system, such a timer can be weeks away. You know that the current minimum time is the soonest you will ever have to go. So once that timer fires, check if it's still needed and act. Otherwise just schedule for next minimal expiration.

That way there is 1 timer, and it usually is set 3 months in the future.

Reply:
> That way there is 1 timer, and it usually is set 3 months in the future.

since these timer (_or scheduled tasks_) do not survive a restart of HA, we need to recreate them on every restart. Those the next expire time could be closer than 90 days, which in turns again needs a check of at least nearest expire date during startup of HA. I agree that it is sufficient to just schedule to the closest expiration 👍

Reply:
with 74d8a9d we now
1) track the next expiration date (_at least now + 90 days if there are no expiring refresh tokens exists_)
2) than remove expired refresh tokens and repeat 1.